bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/tools/DelayedBatchProcessing.java

```diff
@@ -56,18 +56,19 @@ public DelayedBatchProcessing(int delay, Consumer<List<T>> consumer, ScheduledEx
     }
 
     /**
-     * Add new object to the batch process list. If the list was empty, the delay timer
-     * is armed and all successive objects are accumulated from here on.
+     * Add new object to the batch process list. Every time a new object is received,
+     * the delay timer is rescheduled.
      *
      * @param t An object
      */
     @Override
-    public void accept(T t) {
+    synchronized public void accept(T t) {
         queue.add(t);
         final ScheduledFuture<?> scheduledFuture = this.future;
-        if (scheduledFuture == null || scheduledFuture.isDone()) {
-            this.future = executor.schedule(this::run, delay, TimeUnit.MILLISECONDS);
+        if (scheduledFuture != null && !scheduledFuture.isDone()) {
+            scheduledFuture.cancel(true);
         }
```

Comment:
I can see a lot of potential concurrency issues around reassigning the future which mostly has to do with how this interacts with the other methods that cancel the future. I'm going to suggest a couple of changes which will make all of this thread-safe.

First, create a utility method for cancelling the future:
```
private static void cancel(@Nullable Future<?> future){
    Future<?> future = this.future;
    if(future != null){
        future.cancel(false);
    }
}
```
You will notice I don't bother checking if it is done or not since nothing bad will happen if you try to cancel an already finished future.

Then you will want to change 
`protected @Nullable ScheduledFuture<?> future;` 
to 
`protected final AtomicReference<@Nullable Future<?>> futureRef = new AtomicReference<>()`.

Then when you just want to cancel the future you call `cancel(futureRef.getAndSet(null))`
And when you want to schedule the future while canceling the old one you call `cancel(futureRef.getAndSet(executor.schedule(this::run, delay, TimeUnit.MILLISECONDS)));

A few of the other methods would also change slightly.
```
    /**
     * Return true if there is a delayed processing going on.
     */
    public boolean isArmed() {
        ScheduledFuture<?> scheduledFuture = this.futureRef.get();
        return scheduledFuture != null && !scheduledFuture.isDone();
    }
```

Doing all this will allow you to safely remove the `synchronized` you added to `accept` as well as making all sure all of your other code is interacting with your future in an atomic way. I think using an AtomicReference here is more performant and less error prone than trying to basically synchronize every method in this class.

Reply:
Wow, thanks for the detailed instructions @cpmeister, I will add them :+1:  @jochen314, do you have anything to add? Do you agree with the proposed solution? I don't want to touch the DelayedBatchProccesing class without your approval :wink:

Reply:
I am 100% with @cpmeister here.
Personally I would even try a method like
```
private boolean setNewFuture(ScheduledFuture<?> newFuture) {
    Future<?> oldFuture = this.future.getAndSet(newFuture);
    boolean result = true;
    if(future != null){
        result = future.isDone();
        future.cancel(false);
    }
    return result;
}
```

And just one more thing:
There is the possibility that:
1. The future times out
2. We receive a new call to `accept()`

Does your caode then schedule another timeout-future?
I think, this should not be done. That's why my proposed method returns, if the old future was already done, so we could use that, not to trigger a new timeout, once a timeout occured.

Reply:
Hi @jochen314!

First of all, many thanks for your message, it's good to see you here :wink: But... I don't understand what you mean with your last concerns.

1. If the future times out, that means that no new messages have arrived in X seconds, so the ```run``` method is called.
2. A new call to ```accept()``` just reschedules the future, resetting the process.

To make things more clear, the current DelayedBatchProccesing class behavior is:

- The first time the ```accept()``` method is called, the future timer is set and new objects are collected when ```accept()``` is triggered. After the ```delay``` time has passed, the ```run``` method is called.

With the proposed changes, the new behavior will be:

- The first time the ```accept()``` method is called, the future timer is set, but it is rescheduled when a new object is accepted. The ```run``` method will be called only after ```delay``` seconds have passed since the last message.

The new logic ensures that discovery processes that take longer than initially anticipated can be completed successfully, as it was suggested in https://github.com/openhab/openhab-addons/pull/5963#issuecomment-523585015. Perhaps I'm explaining something that is obvious, but I just wanted to avoid any misunderstanding.

Anyway, tomorrow I will update the PR with the changes proposed by @cpmeister. Stay tuned!